File: 20/1.py
from collections import deque

class Solution:
    def isValid(self, s: str) -> bool:
        stack = deque()
        for char in s:
            if char == '(' or char == '{' or char == '[':
                stack.append(char)
            else:
                if not stack: return False
                if char == ')' and stack.pop() != '(': return False
                elif char == '}' and stack.pop() != '{': return False
                elif char == ']' and stack.pop() != '[': return False
            
        return not stack

def main():
    s = input()
    solution = Solution()
    print(solution.isValid(s))

if __name__ == "__main__":
    main()

File: 20/1.cpp
#include <iostream>
#include <string>
#include <stack>

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;
        for (char &c : s) {
            switch (c) {
                case '(':
                case '[':
                case '{':
                    stk.push(c);
                    break;
                case ')':
                    if (stk.empty() || stk.top() != '(') return false; else stk.pop();
                    break;
                case ']':
                    if (stk.empty() || stk.top() != '[') return false; else stk.pop();
                    break;
                case '}':
                    if (stk.empty() || stk.top() != '{') return false; else stk.pop();
                    break;
            }
        }
        return stk.empty();
    }
};

int main() {
    string s;
    cin >> s;
    Solution sol;
    cout << sol.isValid(s) << endl;
    return 0;
}